pkg/handler/flows.go

```diff
@@ -0,0 +1,346 @@
+	hlog.Debugf("getFlows query params: %s", params)
+
+	start, err := getStartTime(params)
+	if err != nil {
+		return nil, http.StatusBadRequest, err
+	}
+	end := params.Get(endTimeKey)
+	limit := params.Get(limitKey)
+	matchAny := params.Get(matchKey) == anyMatchValue
+	reporter := params.Get(reporterKey)
+	rawFilters := params.Get(filtersKey)
+	filters := parseFilters(rawFilters)
+	grouped, err := groupFilters(&cfg, filters, matchAny)
+	if err != nil {
+		return nil, http.StatusBadRequest, err
+	}
+
+	var rawJSON []byte
+	if len(grouped) > 1 {
+		// match any, and multiple filters => run in parallel then aggregate
+		res, code, err := fetchParallel(&cfg, client, grouped, start, end, limit, reporter)
+		if err != nil {
+			return nil, code, errors.New("Error while fetching flows from Loki: " + err.Error())
+		}
+		rawJSON = res
+	} else {
+		// else, run all at once
+		qb := loki.NewFlowQueryBuilder(&cfg, start, end, limit, reporter)
+		if len(grouped) > 0 {
+			err := qb.Filters(grouped[0])
+			if err != nil {
+				return nil, http.StatusBadRequest, err
+			}
+		}
+		query := qb.Build()
+		resp, code, err := fetchSingle(query, client)
+		if err != nil {
+			return nil, code, errors.New("Error while fetching flows from Loki: " + err.Error())
+		}
+		rawJSON = resp
+	}
+
+	hlog.Tracef("GetFlows raw response: %v", rawJSON)
+	return rawJSON, http.StatusOK, nil
+}
+
+func fetchParallel(cfg *loki.Config, lokiClient httpclient.HTTPClient, groupedFilters []map[string]string, start, end, limit, reporter string) ([]byte, int, error) {
+	// Run queries in parallel, then aggregate them
+	resChan := make(chan model.QueryResponse, len(groupedFilters))
+	errChan := make(chan errorWithCode, len(groupedFilters))
+	var wg sync.WaitGroup
+	wg.Add(len(groupedFilters))
+
+	for _, group := range groupedFilters {
+		go func(filters map[string]string) {
+			defer wg.Done()
+			qb := loki.NewFlowQueryBuilder(cfg, start, end, limit, reporter)
+			err := qb.Filters(filters)
+			if err != nil {
+				errChan <- errorWithCode{err: err, code: http.StatusBadRequest}
+				return
+			}
+			query := qb.Build()
+			resp, code, err := fetchSingle(query, lokiClient)
+			if err != nil {
+				errChan <- errorWithCode{err: err, code: code}
+			} else {
+				var qr model.QueryResponse
+				err := json.Unmarshal(resp, &qr)
+				if err != nil {
+					errChan <- errorWithCode{err: err, code: http.StatusInternalServerError}
+				} else {
+					resChan <- qr
+				}
+			}
+		}(group)
+	}
+
+	wg.Wait()
+	close(resChan)
+	close(errChan)
+
+	for errWithCode := range errChan {
+		return nil, errWithCode.code, errWithCode.err
+	}
+
+	// Aggregate results
+	var aggregated model.QueryResponse
+	var aggStreams model.Streams
+	for r := range resChan {
+		if streams, ok := r.Data.Result.(model.Streams); ok {
+			if len(aggStreams) == 0 {
+				aggStreams = streams
+				aggregated = r
+			} else {
+				aggStreams = append(aggStreams, streams...)
```

Comment:
How do you plan to remove duplicates here ?
What about the limit ?

Reply:
good points, it needs dedup, I will work on it.

for limit, either we just throw away results outside of the limit, but but it might be a shame, so another option could be to find some twist in the UI to not present the limit option as something hard-defined, but with a more qualitative wording instead? (e.g 100 => fetch few, 500 => fetch more , 1000 => fetch many more .. or find something better) any idea?

For sure this solution comes with some challenges, but IMO we cannot continue without using loki indexes if we want to scale